File: 10-Adapter/src/go/paypal_api.go
package main

import (
	"fmt"
	"time"
)

// Third-party PayPal API — different interface, amounts as float32.
type PayPalApi struct{}

type PayPalResponse struct {
	StatusCode int // 0 = success
	PaymentId  string
}

func (p *PayPalApi) SendPayment(amount float32, currencyCode string) PayPalResponse {
	fmt.Printf("  [PayPal] Sending %.2f %s\n", amount, currencyCode)
	return PayPalResponse{StatusCode: 0, PaymentId: fmt.Sprintf("PAY-%d", time.Now().UnixNano())}
}

func (p *PayPalApi) ReversePayment(paymentId string) PayPalResponse {
	fmt.Printf("  [PayPal] Reversing payment %s\n", paymentId)
	return PayPalResponse{StatusCode: 0, PaymentId: fmt.Sprintf("REV-%d", time.Now().UnixNano())}
}

File: 10-Adapter/src/rust/paypal_api.rs
/// Third-party PayPal API — different interface, amounts as f32.
pub struct PayPalApi;

pub struct PayPalResponse {
    pub status_code: i32,  // 0 = success
    pub payment_id: String,
}

impl PayPalApi {
    pub fn send_payment(&self, amount: f32, currency_code: &str) -> PayPalResponse {
        println!("  [PayPal] Sending {} {}", amount, currency_code);
        PayPalResponse {
            status_code: 0,
            payment_id: format!("PAY-{}", std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH).unwrap().as_nanos()),
        }
    }

    pub fn reverse_payment(&self, payment_id: &str) -> PayPalResponse {
        println!("  [PayPal] Reversing payment {}", payment_id);
        PayPalResponse {
            status_code: 0,
            payment_id: format!("REV-{}", std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH).unwrap().as_nanos()),
        }
    }
}
